hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,319 @@
+import java.time.LocalDate;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static com.hedera.services.context.properties.StaticPropertiesHolder.STATIC_PROPERTIES;
+import static com.hedera.services.ledger.accounts.staking.RewardCalculator.zoneUTC;
+import static com.hedera.services.ledger.interceptors.StakeChangeManager.finalBalanceGiven;
+import static com.hedera.services.ledger.interceptors.StakeChangeManager.hasStakeFieldChanges;
+import static com.hedera.services.ledger.interceptors.StakeChangeManager.isWithinRange;
+import static com.hedera.services.ledger.properties.AccountProperty.DECLINE_REWARD;
+
+public class StakeAwareAccountsCommitsInterceptor extends AccountsCommitInterceptor {
+	private final StakeChangeManager manager;
+	private final Supplier<MerkleNetworkContext> networkCtx;
+	private final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo;
+	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
+	private final RewardCalculator rewardCalculator;
+	private final SideEffectsTracker sideEffectsTracker;
+	private final GlobalDynamicProperties dynamicProperties;
+
+	private boolean rewardsActivated;
+	private boolean rewardBalanceChanged;
+	private long newRewardBalance;
+
+	private static final Logger log = LogManager.getLogger(StakeAwareAccountsCommitsInterceptor.class);
+	private static final long STAKING_FUNDING_ACCOUNT_NUMBER = 800L;
+
+	public StakeAwareAccountsCommitsInterceptor(
+			final SideEffectsTracker sideEffectsTracker,
+			final Supplier<MerkleNetworkContext> networkCtx,
+			final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo,
+			final GlobalDynamicProperties dynamicProperties,
+			final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
+			final RewardCalculator rewardCalculator,
+			final StakeChangeManager manager) {
+		super(sideEffectsTracker);
+		this.manager = manager;
+		this.networkCtx = networkCtx;
+		this.stakingInfo = stakingInfo;
+		this.accounts = accounts;
+		this.rewardCalculator = rewardCalculator;
+		this.sideEffectsTracker = sideEffectsTracker;
+		this.dynamicProperties = dynamicProperties;
+	}
+
+	@Override
+	public void preview(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		// if the rewards are activated previously they will not be activated again
+		rewardsActivated = rewardsActivated || networkCtx.get().areRewardsActivated();
+		rewardBalanceChanged = false;
+
+		super.preview(pendingChanges);
+
+		// The latest period by which an account must have started staking, if it can be eligible for a
+		// reward; if staking is not active, this will return Long.MIN_VALUE so no account is eligible
+		final var latestEligibleStart = rewardCalculator.latestRewardableStakePeriodStart();
+
+		// boolean to track if the account has been rewarded already with one of the pending changes
+		Set<Long> hasBeenRewarded = new HashSet<>();
```

Comment:
A `boolean[] hasBeenRewarded` array parallel to the `pendingChanges` arrays will be much faster. (Where we begin each iteration of the loop below by setting `hasBeenRewarded[i] = false`.)

Reply:
I thought of set because, when we resize `pendingChanges` we need to resize `hasBeenRewarded` as well , if `hasBeenRewarded = pendingChanges.size()` initially. What is your thought ?